### ISSUE TITLE
"area-screenshot," "area-screenshot-clip" do not work, under any keybinding

### ISSUE THREAD

**Issue by wchargin** [linuxmint/cinnamon#3981]
Since upgrading from 17 Qiana to 17.1 Rebecca (via `mintupdate`), the <kbd>Shift</kbd><kbd>PrintScreen</kbd> keybinding no longer works for me, nor does <kbd>Shift</kbd><kbd>Ctrl</kbd><kbd>PrintScreen</kbd>.

I checked in `cinnamon-settings` &rarr; Keyboard &rarr; Keyboard Shortcuts &rarr; System &rarr; Screenshots and Recording, and the keys are properly bound. If I add another binding for the same operation, it still doesn't work.

Running `gnome-screenshot -a` from a terminal gives the expected behavior.

I have run `cinnamon-desktop-migrate-mediakeys`. I don't know if it was working before I did this.

**Comment by WChargin** [comment]
Any thoughts? I'd be happy to work on a patch if someone could show me where to start. Thanks!

**Comment by nutztherookie** [comment]
i can confirm this problem.
also i've never run `cinnamon-desktop-migrate-mediakeys` (knowingly)  and still it's not working

**Comment by juanluperez** [comment]
Same problem here in a new Linux Mint 17.3 intall.

**Comment by juanluperez** [comment]
There's a trick a I do to make it work. I keep Ctrl+Shit+Prnt.Screen pushed for few seconds and when released it works. It works in a good probability.

**Comment by WChargin** [comment]
@juanluperez: whoa, can confirm—that's mind-boggling. any ideas why?

**Comment by porpoisepor** [comment]
Stumbled upon this using a fresh Mint 17.3 install too.
This has been described here https://forums.linuxmint.com/viewtopic.php?f=60&t=210288 and also here https://forums.linuxmint.com/viewtopic.php?t=202909

tl;dr:
Two separate workarounds are:
1) 
> holding Shift and doing a quick tap of the PrtScrn works everytime

2) Using a script with sleep before running the screenshot command
```
#!/bin/sh

sleep 0.125
gnome-screenshot --area
```

**Comment by anandrkris** [comment]
If I hold the keys for a while, I can get it to work. 
MATE seems to be affected as well.
https://github.com/mate-desktop/mate-utils/issues/37

**Comment by ManIVIctorious** [comment]
I just tried all screenshot/recording shortcuts and they all seem to work for me on Arch Linux with cinnamon 3.0.7 and gnome-screenshot 3.20.1 installed.
Is this still an issue for you with latest cinnamon?

**Comment by juanluperez** [comment]
Yes, this is still a problem with new cinnamon. Nothing changes.

**Comment by christianbueno1** [comment]
Hello friends, I just had installed Antergos [Arch Linux]+ Cinnamon , I has to say there is a problem for take area screenshot.
Similar to my friend @anandrkris , I has to hold the keys for a while.
cinnamon version: 3.0.7
Linux Kernel: 4.8.6-1ARCH 
Greetings from Ecuador

**Comment by christianbueno1** [comment]
Thank s to [Joriktos.](https://forums.linuxmint.com/viewtopic.php?f=90&t=202909&p=1235957#p1235957)
First let's go to delete the already existing keybinding. Go to 
>System Settings>Keyboard>Shortcuts

In the left panel select
>System>Screenshot and Recording

Click in Take a screenshot of an area, below in Keyboard bindings section double click in Shift+Print and BackSpace for clear the existing keybinding. After that, go below and click 
>Add custom shortcut

type a name "New_area_shortcut" for example, and in command type 
>sh -c 'sleep 0.125; gnome-screenshot -a'

and after click Add.

OS: Antergos [Arch Linux]
cinnamon version: 3.0.7
Linux Kernel: 4.8.6-1ARCH
gnome-screenshot 3.22.0

**Comment by ManIVIctorious** [comment]
I have realized that cinnamon sometimes has to be restarted for new shortcuts to work, could you please retry after a cinnamon restart?

**Comment by JerryGreen** [comment]
Almost a year behind and still no fix :{

Btw thanks for hint with holding

**Comment by margamanterola** [comment]
This is still the case on Debian stretch, running cinnamon 3.2

I can confirm the information that is mentioned in #5263: the problem depends on how long you press the PrtScn key.  If you do a very short keypress, it works fine.  Doing a longer (normal) keypress, doesn't work.

Looking at .xsession-errors you can see that in the second case (normal length keypress), the screenshot was already triggered, and that the are had width = 0 and height = 0

**Comment by margamanterola** [comment]
I suspect this is actually a long-standing bug in gnome-screenshot.  It's been reported for mate-screenshot:
https://github.com/mate-desktop/mate-utils/issues/37

I think this upstream bug is reporting about the same issue, although it doesn't have much info:
https://bugzilla.gnome.org/show_bug.cgi?id=682626

**Comment by Deleetdk** [comment]
Long standing bug, still not fixed in Mint 18.2.

- https://bugs.launchpad.net/linuxmint/+bug/1620318
- https://forums.linuxmint.com/viewtopic.php?t=202909

Sigh!

**Comment by soomtong** [comment]
I have 3 Mint 18.2 machines.
the 2 of them work well with any print screen key.
but my main desktop does not work as a normal procedure.

anyway there is a bug, we still know another solution.
cheers!

**Comment by Marinofull** [comment]
I came here hope for a solution, but I'm in Mint 18.3 and the bug still remains ;(

**Comment by smurphos** [comment]
I remapped to not use my PrintScr button - Crtl-S - full screen screenshot, Super-S - Current Window. All is well.

**Comment by adenbley** [comment]
Mint 19 still has issue. Thanks for the tips though.

**Comment by Sraw** [comment]
I can confirm this is still a bug in Mint 19. Hope it can be solved. I have to double press Print to get it work.

**Comment by lucidyan** [comment]
Still a bug in Mint 19

**Comment by davmillar** [comment]
Can confirm this also occurs when installing Cinnamon on an existing Fedora 28 machine. It works fine when I log into Gnome 3 instead.

**Comment by Sraw** [comment]
@clefebvre Do you mean update `gnome-screenshot` to `3.25.0`? I've tested it, nothing improved. Still a bug. Sometimes works, sometimes not.

**Comment by mtwebster** [comment]
Did you try the package he linked?  Upstream 3.25 does not contain this fix (it has been submitted but not yet accepted).

As to your other question, most likely nothing changed with gnome-screenshot itself.  Hundreds of packages surrounding and being used by gnome-screenshot did.  The way gnome-screenshot works it likely ended up being affected by what is more or less behavioral drift by the desktop and other components of the os over the years.

**Comment by Sraw** [comment]
Now I have tried it. Whoa, it works seamlessly. At least for now.

Thank you @mtwebster and @clefebvre .

**Comment by altarin** [comment]
> Stumbled upon this using a fresh Mint 17.3 install too.
> This has been described here https://forums.linuxmint.com/viewtopic.php?f=60&t=210288 and also here https://forums.linuxmint.com/viewtopic.php?t=202909
> 
> tl;dr:
> Two separate workarounds are:
> 1)
> 
> > holding Shift and doing a quick tap of the PrtScrn works everytime
> 
> 1. Using a script for the ShiftPrtScn shortcut with sleep before running the screenshot command:
> 
> ```
> #!/bin/sh
> 
> sleep 0.125
> gnome-screenshot --area
> ```

For me works, but with `sleep 0.25`

**Comment by altarin** [comment]
> Thanks to [Joriktos.](https://forums.linuxmint.com/viewtopic.php?f=90&t=202909&p=1235957#p1235957)
> First let's go to delete the already existing keybinding. Go to
> 
> > System Settings>Keyboard>Shortcuts
> 
> In the left panel select
> 
> > System>Screenshot and Recording
> 
> Click in Take a screenshot of an area, below in Keyboard bindings section double click in Shift+Print and BackSpace for clear the existing keybinding. After that, go below and click
> 
> > Add custom shortcut
> 
> type a name "New_area_shortcut" for example, and in Command area type
> 
> > sh -c 'sleep 0.125; gnome-screenshot -a'
> 
> and after click Add.
> 
> OS: Antergos [Arch Linux]
> cinnamon version: 3.0.7
> Linux Kernel: 4.8.6-1ARCH
> gnome-screenshot 3.22.0

For me works, but with `sleep 0.25`

**Comment by LinuxOnTheDesktop** [comment]
It seems to me that this problem persists. Am I right?

**Comment by areographe** [comment]
This problem is still occurring. With the keybinding I was using, pressing <kbd>Shift</kbd><kbd>PrtSc</kbd> would always create a `gnome-screenshot` process, but 4 out of 5 times the process would just fail to work and then stay sleeping until I ran `killall gnome-screenshot`. Changing to a custom shortcut with `sleep 0.25` beforehand still works, though.

**Comment by Deleetdk** [comment]
I will donate 100 USD to whoever submits an accepted pull request solution for this. It's been 5 years now!

**Comment by areographe** [comment]
Agreed @Deleetdk, I'm skint at the moment but will definitely throw a few USD towards a solution if one is found. I should just mention, when @juanluperez says:

> There's a trick a I do to get it working. I keep Ctrl+Shift+Prnt.Screen pushed for few seconds and when released it works. It works in a good probability.

In my experience, opening `htop` and then testing the "keep [the key combination] pushed for a few seconds" technique shows that many `gnome-screenshot` processes appear, the longer the keys are pressed the more identical processes appear, and when the crosshairs finally appear and I select the area and successfully take the screenshot, one `gnome-screenshot` process disappears, but all the previous ones spawned during holding down the keys remain, again until I run `killall gnome-screenshot`.

**Comment by voleresistor** [comment]
> In my experience, opening `htop` and then testing the "keep [the key combination] pushed for a few seconds" technique shows that many `gnome-screenshot` processes appear, the longer the keys are pressed the more identical processes appear, and when the crosshairs finally appear and I select the area and successfully take the screenshot, one `gnome-screenshot` process disappears, but all the previous ones spawned during holding down the keys remain, again until I run `killall gnome-screenshot`.

+1 more for this. Holding the key opens many instances of `gnome-screenshot` for me on Fedora 32. These do not exit unless you kill the processes and the buildup seems to eventually prevent the hotkey from working even by using the hold [key combo] "workaround" until you kill them.

The custom shortcut with
>sh -c 'sleep 0.125; gnome-screenshot -a'

DOES work on the first keypress every time and doesn't seem to result in spawning multiple additional processes so thank you to whoever initially figured that out.

**Comment by mtwebster** [comment]
I'm going to call this tentatively fixed by this:https://github.com/linuxmint/cinnamon-settings-daemon/commit/7daaa5c58181430ef7d8b63b73c2a5a8d8d4a442

We've been patching gnome-screenshot on mint for a while now to work around the issue with area screenshots, but this should allow any version of gnome-screenshot to work properly in Cinnamon.

I had submitted a patch upstream but it was never really looked at, so this is about the next best we can probably do.

**Comment by dandv** [comment]
I know nothing about Mint's internals. I'm curious why screenshot taking is so complex as to lead to [this many bugs](https://github.com/linuxmint/cinnamon/issues?q=in%3Atitle+screenshot)?

**Comment by LinuxOnTheDesktop** [comment]
> I'm going to call this tentatively fixed by this:[linuxmint/cinnamon-settings-daemon@7daaa5c](https://github.com/linuxmint/cinnamon-settings-daemon/commit/7daaa5c58181430ef7d8b63b73c2a5a8d8d4a442)
> 
> We've been patching gnome-screenshot on mint for a while now to work around the issue with area screenshots, but this should allow any version of gnome-screenshot to work properly in Cinnamon.
> 
> I had submitted a patch upstream but it was never really looked at, so this is about the next best we can probably do.

@mtwebster 

Might Mint be able to do any better, on this front, for Mint 20.1, please?

**Comment by mtwebster** [comment]
I'm pretty satisfied with that fix I linked.

The issue, simplified, is that the mechanism for stretching the rectangle over what you want to take a screenshot of requires a 'grab' of the pointer/keyboard - meaning, it wants to acquire temporary input focus over the rest of the desktop.

However, the way system-wide shortcuts (like the screenshot hotkeys) *also* take a brief grab of the focus, and it hasn't relinquished that grab yet by the time gnome-screenshot wants it. So gnome-screenshot takes that as failure and aborts. You could sometimes make it work if you tried enough times, you could break thru this 'race' condition.

When we patched gnome-screenshot, we made it attempt to get this grab multiple times until it got it (with a short delay in between attempts). This worked great but it only worked in Mint. The cinnamon-settings-daemon patch addresses this in the same way,  it's just done in a spot that will allow the fix to work universally now, on any distro (and we can stop patching a foreign package).

**Comment by foxel** [comment]
My slight variation of the fix, maybe usable for those who get there googling for a solution.

put the following into *~/.local/bin/gnome-screenshot*

```shell
#!/bin/sh

sleep 0.125

LANGUAGE=en /usr/bin/gnome-screenshot $@
```
